File: src/com/alchitry/labs/parsers/types/PinConstraint.java
package com.alchitry.labs.parsers.types;

import com.alchitry.labs.hardware.boards.Board;
import com.alchitry.labs.hardware.pinout.PinConverter;

public class PinConstraint {
	private String pin;
	private String port;
	private int portBit;
	private boolean isMultiBit;
	private boolean pullUp;
	private boolean pullDown;

	public PinConstraint(String pin, String port) {
		this(pin, port, 0, false, false, false);
	}

	public PinConstraint(String pin, String port, int portBit, boolean isMultiBit, boolean pullUp, boolean pullDown) {
		this.pin = pin;
		this.port = port;
		this.portBit = portBit;
		this.isMultiBit = isMultiBit;
		this.pullUp = pullUp;
		this.pullDown = pullDown;
	}
	
	public String getPin() {
		return pin;
	}
	
	public String getPort() {
		return port;
	}
	
	public String getBoardConstraint(Board board) {
		StringBuilder sb = new StringBuilder();
		PinConverter pc = board.getPinConverter();
		
		String portName = port + (isMultiBit ? ("[" + portBit + "]") : "");
		sb.append("# ").append(portName).append(" => ").append(pin).append(System.lineSeparator());
		
		if (board.isType(Board.AU)) {
			sb.append("set_property PACKAGE_PIN ").append(pc.getFPGAPin(pin)).append(" [get_ports {").append(portName).append("}]").append(System.lineSeparator());
			sb.append("set_property IOSTANDARD LVCMOS33 [get_ports {").append(portName).append("}]").append(System.lineSeparator());
			if (pullUp)
				sb.append("set_property PULLUP true [get_ports {").append(portName).append("}]").append(System.lineSeparator());
			if (pullDown)
				sb.append("set_property PULLDOWN true [get_ports {").append(portName).append("}]").append(System.lineSeparator());
		} else if (board.isType(Board.CU)) {
			sb.append("set_io ").append(portName).append(" ").append(pc.getFPGAPin(pin));
			if (pullUp)
				sb.append(" -pullup yes");
			sb.append(System.lineSeparator());
		}
		
		return sb.toString();
	}
}

File: src/com/alchitry/labs/parsers/types/PinConstraint.kt
package com.alchitry.labs.parsers.types

import com.alchitry.labs.hardware.boards.Board

class PinConstraint @JvmOverloads constructor(val pin: String, val port: String, private val portBit: Int = 0, private val isMultiBit: Boolean = false, private val pullUp: Boolean = false, private val pullDown: Boolean = false) {
    fun getBoardConstraint(board: Board): String {
        val sb = StringBuilder()
        val pc = board.pinConverter
        val portName = port + if (isMultiBit) "[$portBit]" else ""
        when {
            board.isType(Board.AU or Board.AU_PLUS) -> {
                sb.append("set_property PACKAGE_PIN ").append(pc.getFPGAPin(pin)).append(" [get_ports {").append(portName).append("}]").append(System.lineSeparator())
                sb.append("set_property IOSTANDARD LVCMOS33 [get_ports {").append(portName).append("}]").append(System.lineSeparator())
                if (pullUp) sb.append("set_property PULLUP true [get_ports {").append(portName).append("}]").append(System.lineSeparator())
                if (pullDown) sb.append("set_property PULLDOWN true [get_ports {").append(portName).append("}]").append(System.lineSeparator())
            }
            board.isType(Board.CU) -> {
                sb.append("set_io ").append(portName).append(" ").append(pc.getFPGAPin(pin))
                if (pullUp) sb.append(" -pullup yes")
                sb.append(System.lineSeparator())
            }
            else -> {
                error("Unsupported board type")
            }
        }
        return sb.toString()
    }
}